src/file_pipe_log/file_mgr.rs

```diff
@@ -0,0 +1,660 @@
+                self.first_seq = file_seq;
+                purged_file_list
+            } else {
+                FileList::new(0, VecDeque::default())
+            };
+        (
+            FileState {
+                first_seq: self.first_seq,
+                total_len: self.fds.len(),
+            },
+            splitted_list,
+        )
+    }
+}
+
+/// A collection of files for managing all log files.
+///
+/// Log files consist of two parts:
+/// - `Stale` part, obsolete files but temporarily stashed for recycling.
+/// - `Active` part, active files for accessing.
+pub struct FileCollection<F: FileSystem> {
+    file_system: Arc<F>,
+    queue: LogQueue,
+    paths: Paths,
+    file_format: LogFileFormat,
+    target_file_size: usize,
+    capacity: usize,
+    /// File list to collect all active files
+    active_files: CachePadded<RwLock<FileList<F>>>,
+    /// File list to collect all stale files
+    stale_files: CachePadded<RwLock<FileList<F>>>,
+}
+
+impl<F: FileSystem> FileCollection<F> {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        file_system: Arc<F>,
+        queue: LogQueue,
+        paths: Paths,
+        file_format: LogFileFormat,
+        target_file_size: usize,
+        capacity: usize,
+        active_files: FileList<F>,
+        stale_files: FileList<F>,
+    ) -> Self {
+        Self {
+            file_system,
+            queue,
+            paths,
+            file_format,
+            target_file_size,
+            capacity,
+            active_files: CachePadded::new(RwLock::new(active_files)),
+            stale_files: CachePadded::new(RwLock::new(stale_files)),
+        }
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.active_files.read().len() + self.stale_files.read().len()
+    }
+
+    #[inline]
+    pub fn get_fd(&self, file_seq: FileSeq) -> Result<Arc<F::Handle>> {
+        self.active_files.read().get(file_seq)
+    }
+
+    #[inline]
+    pub fn back(&self) -> Option<FileWithFormat<F>> {
+        self.active_files.read().back().map(|fd| FileWithFormat {
+            handle: fd.handle.clone(),
+            ..*fd
+        })
+    }
+
+    #[inline]
+    pub fn active_file_span(&self) -> Option<(FileSeq, FileSeq)> {
+        self.active_files.read().span()
+    }
+
+    #[inline]
+    #[cfg(test)]
+    fn stale_file_span(&self) -> Option<(FileSeq, FileSeq)> {
+        self.stale_files.read().span()
+    }
+
+    #[inline]
+    pub fn push_back(&self, seq: FileSeq, handle: FileWithFormat<F>) -> FileState {
+        self.active_files.write().push_back(seq, handle)
+    }
+
+    #[inline]
+    pub fn get_target_file_size(&self) -> usize {
+        self.target_file_size
+    }
+
+    /// Rotate a new file handle and return it to the caller.
+    ///
+    /// Returns the relevant `FileSeq`, and `FileWithFormat` of the file
+    /// handle to the caller.
+    /// Attention please, the returned fd should be manually appended to current
+    /// `FileCollection`.
```

Comment:
Maybe move the file writer creation inside, so that the whole step can be done atomically here.